### PR TITLE
Detect prompt-cache breaks and stop doomed compaction loops

### PR DESCRIPTION
**Pull request by ilblackdragon**
**Stack 1/4 on #5959.** Lessons from studying Claude Code's execution engine, applied to the measured KV-cache collapse (82% hit rate at ~120 calls → 29% at 425 calls on claw-swe-bench run `799636ce`, a ~3.5× cost multiplier on long agentic turns).

## Part 1 — prompt-cache break detector (`ironclaw_runner::model_gateway`)

Every completed model call now emits a per-call `tracing::debug!` cache-usage series (cache_read / cache_creation / input tokens), and a **`tracing::warn!` "prompt cache break detected"** when the turn-over-turn cache read drops >5% AND >10K tokens — with attribution hints: seconds since last call, whether the tool-definition list changed (name-list hash), whether the system prompt changed (content hash). Per-run scoping via a bounded activity log on both gateway flavors. This is the diagnostic that tells us *why* the hit rate collapses (compaction rewrite vs provider routing vs surface churn) instead of guessing from run totals.

## Part 2 — compaction circuit breaker (`ironclaw_agent_loop`)

Claude Code added the same guard after measuring ~250K wasted API calls/day from doomed auto-compaction loops. If compaction runs 3 consecutive times without bringing the token estimate back under the trigger threshold, the circuit opens for the rest of the run (one warn, covers forced/byte-cap compaction too; old checkpoints rehydrate with the breaker closed). State lives in the typed `CompactionStrategyState` slot per the strategies contract; no default parameter values changed. Family fingerprints gained `ineffective_trip_limit=3` and digests were regenerated red→green.

## Testing

Red-then-green throughout: pure-function tests for the break predicate (boundary cases), activity-log scope isolation + attribution flags, slot transition tests, strategy skip tests, a through-the-caller `PromptStage` executor test, and a gateway-tier `#[traced_test]` asserting the warn fires within a run (and stays quiet at exact thresholds). `cargo test -p ironclaw_agent_loop` fully green in seconds (this stack's base also pauses tokio time in the scenario test targets, which the deep availability backoff from #5959 had slowed to ~423s). fmt + clippy clean.

Deferred (documented in code): message-prefix divergence attribution (prior-call messages aren't retained at the gateway), tool *schema* hashing (names+count only), eviction-on-run-completion (no lifecycle signal at this layer).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/ironclaw_agent_loop/src/executor/prompt.rs [hunk 1]
```diff
@@ -586,6 +587,31 @@ impl<'a, 'b> PromptCompactionStep<'a, 'b> {
...
+        // estimate is still at/above the trigger threshold would fire again
+        // immediately — it didn't help. The retained estimate excludes the
+        // injected summary's tokens (the rebuilt bundle isn't known yet), so
+        // this under-counts and only makes the breaker harder to trip. After
+        // `INEFFECTIVE_COMPACTION_TRIP_LIMIT` consecutive ineffective runs the
+        // strategies stop compacting for the remainder of the run — Claude
+        // Code measured ~250K wasted API calls/day from exactly this
+        // compact-recompact doom loop before adding a breaker.
+        let circuit_was_open = state.compaction_state.compaction_circuit_open;
+        state.compaction_state = state
```
Comment (serrrfirat) [inline crates/ironclaw_agent_loop/src/executor/prompt.rs]: **[High · Performance/Correctness · 96%] Circuit breaker measures before the summary is added**

The breaker classifies compaction immediately after `retain_after_sequence`, when the estimate contains only the retained tail and explicitly excludes the newly injected summary. A compactor can return a summary large enough that the rebuilt prompt stays above threshold, yet every attempt is recorded as effective because the tail alone is below threshold. The next iteration compacts again and resets the counter, so the doomed loop can continue through the 1,024-iteration backstop.

This was also flagged by the Tests lens: the caller test seeds the counter at `limit - 1` and never proves three real prompt-stage compactions trip the breaker.

**Fix:** Measure after rebuilding the bundle including the summary, or carry a pending observation into the next iteration and compare the refreshed total before incrementing/resetting. Add a caller test driving three actual ineffective compactions and suppressing the fourth.

File: crates/ironclaw_agent_loop/src/strategies/compaction.rs [hunk 2]
```diff
@@ -60,6 +68,7 @@ impl DefaultCompactionStrategy {
             drop_through_seq,
             preserve_tail_tokens: self.preserve_tail_tokens,
             deadline_ms: self.deadline_ms,
+            trigger_threshold_tokens: self.prompt_context_budget.visible_transcript_tokens(),
```
Comment (serrrfirat) [inline crates/ironclaw_agent_loop/src/strategies/compaction.rs]: **[High · Bug · 95%] Forced compactions use the wrong effectiveness threshold**

Every trigger carries the normal visible-transcript threshold, including `force_compact_on_next_iteration` raised by `ByteCapStrategy`. A forced compaction can be completely ineffective while the retained prompt remains below that unrelated threshold; repeated ~32K capability-result overflows under the default ~108K transcript threshold will reset the counter rather than advance it. This contradicts the stated requirement that the breaker cover forced/byte-cap compactions.

**Fix:** Carry trigger-specific effectiveness data. For forced compactions compare against pre-compaction size or a byte-cap-specific target; keep the transcript threshold only for threshold-triggered compaction.

File: crates/ironclaw_runner/src/model_gateway.rs [hunk 3]
```diff
@@ -1169,6 +1413,13 @@ where
                     return Err(map_provider_error(error));
                 }
             };
+            if let Some(scope) = cache_scope.as_ref() {
```
Comment (serrrfirat) [inline crates/ironclaw_runner/src/model_gateway.rs:1199]: **[Medium · Tests · 90%] Tool-capable cache telemetry path is untested**

The gateway-level cache-break test exercises only the text `CompletionResponse` branch. No integration test drives `stream_model_with_capabilities` through this distinct `ToolCompletionResponse` branch, so tool-definition hashing, recording, and attribution can regress unnoticed.

**Fix:** Add `tests::llm_gateway::gateway_warns_and_attributes_prompt_cache_break_on_tool_capable_calls`, using two same-run capability calls with a cache collapse and changed tool surface.

File: crates/ironclaw_runner/src/model_gateway.rs [hunk 4]
```diff
@@ -1234,6 +1485,13 @@ where
                             return Err(map_provider_error(error));
                         }
                     };
+                    if let Some(scope) = cache_scope.as_ref() {
```
Comment (serrrfirat) [inline crates/ironclaw_runner/src/model_gateway.rs:1271]: **[Medium · Tests · 85%] Repair retry cache recording has no test**

A repairable tool-output failure performs a second provider call and records it as a separate cache observation, but existing repair tests do not assert this telemetry path.

**Fix:** Add `tests::llm_gateway::gateway_records_prompt_cache_usage_for_tool_repair_retry`, asserting both calls share the run scope and the retry emits a break on a large cache-read drop.

File: crates/ironclaw_runner/src/model_gateway.rs [hunk 5]
```diff
@@ -1079,9 +1105,224 @@ impl CompletionStreamSink for ProviderStreamSink {
...
+            .unwrap_or_else(|poisoned| poisoned.into_inner());
+        let observation = match scopes.get(&run_id) {
+            None => PromptCacheObservation::FirstCall,
+            Some(previous)
+                if is_prompt_cache_break(
+                    previous.cache_read_input_tokens,
+                    usage.cache_read_input_tokens,
+                ) =>
+            {
+                PromptCacheObservation::Break {
```
Comment (serrrfirat) [inline crates/ironclaw_runner/src/model_gateway.rs]: **[Medium · Tests · 80%] System-prompt attribution branch is untested**

Tests cover changed tools with an unchanged system prompt, but not the complementary branch. `system_prompt_changed` could remain permanently false without detection.

**Fix:** Add `tests::model_gateway::prompt_cache_activity_log_attributes_system_prompt_only_break` with unchanged tools and a changed system prompt.

File: crates/ironclaw_runner/src/model_gateway.rs [hunk 6]
```diff
@@ -1079,9 +1105,224 @@ impl CompletionStreamSink for ProviderStreamSink {
...
+            input_tokens = usage.input_tokens,
+            "reborn model gateway prompt cache usage"
+        );
+        if let PromptCacheObservation::Break {
+            previous_cache_read_tokens,
+            tool_definitions_changed,
+            system_prompt_changed,
+        } = observation
+        {
+            tracing::warn!(
```
Comment (serrrfirat) [inline crates/ironclaw_runner/src/model_gateway.rs]: **[Medium · Conventions · 100%] Internal cache diagnostics must not log at warn**

This is internal prompt-cache diagnostic attribution, but `tracing::warn!` appears in and corrupts the REPL/TUI. The repository logging rule directs internal trace analysis to `debug!` (`CLAUDE.md:84`).

**Fix:** Emit at `debug!`, or use a diagnostics-only target excluded from interactive output.

File: crates/ironclaw_agent_loop/src/executor/prompt.rs [hunk 7]
```diff
@@ -586,6 +587,31 @@ impl<'a, 'b> PromptCompactionStep<'a, 'b> {
...
+            .compaction_state
+            .with_compaction_effectiveness_observed(
+                state.compaction_prompt.observed_prompt_tokens,
+                trigger_threshold_tokens,
+            );
+        if !circuit_was_open && state.compaction_state.compaction_circuit_open {
+            tracing::warn!(
```
Comment (serrrfirat) [inline crates/ironclaw_agent_loop/src/executor/prompt.rs]: **[Medium · Conventions · 100%] Compaction breaker diagnostics must not log at warn**

Opening an internal circuit breaker emits `tracing::warn!`, violating the same repository rule that internal engine diagnostics belong at `debug!`; the structured progress path already carries lifecycle information.

**Fix:** Change this diagnostic to `debug!`, or route it through a diagnostics-only target.

File: crates/ironclaw_runner/src/model_gateway.rs [hunk 8]
```diff
@@ -1079,9 +1105,224 @@ impl CompletionStreamSink for ProviderStreamSink {
     }
 }
 
+/// Relative drop factor for cache-break detection: the current call must read
```
Comment (serrrfirat) [inline crates/ironclaw_runner/src/model_gateway.rs]: **[Medium · Thermo/Conventions · 90%] Extract cache telemetry from the 3.3K-line gateway**

This adds roughly 350 lines of cache-break policy, state, hashing, eviction, logging, and tests directly to `model_gateway.rs`, growing it from 2,948 to 3,296 lines. Prompt-cache activity is a cohesive new runtime-adapter concern, and `crates/ironclaw_runner/AGENTS.md` explicitly says to add a new file for a new concern. This makes an existing god-file materially broader.

**Fix:** Move the tracker, observation types, thresholds, signatures, eviction, logging, and unit tests into a private `prompt_cache_activity` submodule; leave only scope construction and recording calls in the gateway.

File: crates/ironclaw_runner/src/model_gateway.rs [hunk 9]
```diff
@@ -1079,9 +1105,224 @@ impl CompletionStreamSink for ProviderStreamSink {
...
+    },
+}
+
+/// Per-run prompt-cache continuity tracker shared by all calls through one
+/// gateway. Benchmarks showed the provider KV cache hit rate collapsing from
+/// ~82% to 29% past ~200 model calls per run (~3.5x input cost); this log
+/// detects each break as it happens and attributes it to cheap request-shape
+/// signals (tool surface or system prompt changed).
+#[derive(Debug, Default)]
+pub struct PromptCacheActivityLog {
```
Comment (serrrfirat) [inline crates/ironclaw_runner/src/model_gateway.rs]: **[Medium · Conventions · 95%] Keep the cache activity log private**

`PromptCacheActivityLog` is `pub`, but every use and all methods are internal to `model_gateway.rs`; related state and observation types are private. This expands the downstream API without providing a usable external contract, and the PR gives no rationale for exporting it.

**Fix:** Remove `pub`, or use `pub(crate)` only if the extracted internal module requires crate-level visibility.

**Review by serrrfirat — CHANGES_REQUESTED** [review]
## Multi-agent + thermo-nuclear review

**Disposition:** Request changes

Five specialist reviewers, an intent analyzer, and an independent thermo-nuclear structural pass reviewed stack PR #5975 against base #5959. After confidence filtering and overlap deduplication, 9 findings remain.

- Security: 0
- Bugs: 1
- Performance/concurrency: 1
- Tests: 4 raw, 3 standalone (one overlaps the core breaker bug)
- Conventions/contracts: 4
- Thermo-nuclear: 1 structural finding, deduplicated with conventions
- Reviewer failures: 0

The two blocking correctness findings show that the circuit breaker can miss both summary-driven and forced/byte-cap compact-recompact loops. The structural pass also rejects adding another cohesive subsystem directly to the 3.3K-line gateway.

**Comment by ilblackdragon** [comment]
All nine findings addressed across three commits (branch restacked on the updated #5959):

**Breaker vs forced compactions (ironloopai blocking) — fixed** (`fea8159a1`). `can_evaluate` now honors `force_compact_on_next_iteration` *before* consulting the breaker, so ContextOverflow `ShrinkContext` recovery and byte-cap overflow always compact; the breaker only gates automatic threshold-triggered compaction. The two tests that pinned the suppressing behavior are replaced with forced-bypass pins, including an executor-tier test driving the prompt stage with the breaker open + force flag set.

**Effectiveness measured before the summary lands (serrrfirat High·96%) — fixed** (`fea8159a1`). The observation is now deferred: the trigger's baseline is stashed in a serialized `pending_effectiveness_baseline` slot and compared against the *next* refreshed prompt (which includes the injected summary) before the counter increments/resets. The old test that seeded `consecutive_ineffective_compactions = LIMIT - 1` now drives three *real* ineffective compactions (retained tail below threshold, summary pushing the rebuild back over — the exact doom loop you described; the old code fails this test) and asserts the fourth is suppressed.

**Forced compactions judged against the wrong threshold (serrrfirat High·95%) — fixed** (`fea8159a1`). `CompactionDecision::Trigger` now carries a `CompactionEffectivenessBaseline` enum: `TriggerThresholdTokens` for automatic triggers, `PreCompactionPromptTokens` for forced ones — a forced compaction is effective iff the prompt actually shrank, so repeated sub-threshold byte-cap overflows advance the breaker instead of resetting it. Pinned at slots tier and executor tier.

**Test gaps (N1–N3) — added** (`39e15ba8d`): tool-capable cache recording through `stream_model_with_capabilities` (two same-run calls, 200K→50K collapse, changed tool surface, asserts `tool_definitions_changed`); repair-retry cache assertions folded into the existing `gateway_repairs_oversized_provider_tool_arguments` test; `system_prompt_changed: true` attribution case added to the classification test.

**Log levels (N4/N5) — downgraded to `debug!`** per the repo logging rule (`39e15ba8d`, `fea8159a1`); the integration test is renamed `gateway_records_prompt_cache_break_within_a_run` and now *fails* if the record line carries WARN/ERROR. The cache-break signal still reaches operators via the structured observation records.

**Structure (N6/N7) — extracted** (`b67a6712c`): ~330 lines of telemetry (tracker, observation types, thresholds, signatures, eviction, unit tests) moved to a private `model_gateway/prompt_cache_activity.rs` submodule, everything `pub(super)` (tighter than pub(crate)); the gateway keeps only scope construction and `record` calls.

**Comment by ilblackdragon** [comment]
@serrrfirat CI green across the stack; all 9 findings addressed (detailed response above) — please re-review.

**Review by serrrfirat — DISMISSED** [review]
